File: org.palladiosimulator.blockchainsystems.threesim/src/main/java/org/palladiosimulator/blockchainsysttems/threesim/creation/connectedsubgraphnetwork/ConnectedSubgraphNetworkGlobalResourcePowerCalculator.java
package org.palladiosimulator.blockchainsysttems.threesim.creation.connectedsubgraphnetwork;

import org.palladiosimulator.blockchainsystems.bscm.blockchainsystemComponentRepository.MiningProcessComponent;
import org.palladiosimulator.blockchainsystems.bscm.nodeallocation.NodeAllocation;
import org.palladiosimulator.blockchainsystems.bscm.p2pnetwork.ConnectedSubgraphsNetworkTopology;
import org.palladiosimulator.blockchainsystems.bscm.p2pnetwork.SubgraphSpecification;
import org.palladiosimulator.blockchainsystems.threesim.creation.abstractions.GlobalResourcePowerCalculator;

/**
 * This class calculates the global resource power of a connected subgraph network topology.
 *
 * @author Yannik Sproll
 */
public class ConnectedSubgraphNetworkGlobalResourcePowerCalculator implements GlobalResourcePowerCalculator {

    private final double _globalResourcePower;

    public ConnectedSubgraphNetworkGlobalResourcePowerCalculator(ConnectedSubgraphsNetworkTopology connectedSubgraphsTopology) {

        _globalResourcePower = connectedSubgraphsTopology
                .getSubgraphs()
                .stream()
                .mapToDouble(x -> calculateResourcePowerOfSubgraph(x))
                .sum();
    }

    private double calculateResourcePowerOfSubgraph(SubgraphSpecification subgraph) {
        return subgraph
                .getNodeTemplates()
                .stream()
                .mapToDouble(x -> getResourcePowerOfAllocation(x.getAllocation()) * x.getNumberOfNodeOccurences())
                .sum();
    }

    private static double getResourcePowerOfAllocation(NodeAllocation nodeAllocation) {
        return nodeAllocation
                .getAllocationContexts()
                .stream()
                .filter(y -> y.getAssemblyContext().getEncapsulatedComponent() instanceof MiningProcessComponent)
                .mapToDouble(y -> y.getResourceContainer().getResourcePower())
                .sum();
    }

    @Override
    public double calculateGlobalResourcePower() {
        return _globalResourcePower;
    }

}

File: org.palladiosimulator.blockchainsystems.threesim/src/main/kotlin/org/palladiosimulator/blockchainsystems/threesim/creation/connectedsubgraphnetwork/ConnectedSubgraphNetworkGlobalResourcePowerCalculator.kt
package org.palladiosimulator.blockchainsystems.threesim.creation.connectedsubgraphnetwork

import org.palladiosimulator.blockchainsystems.bscm.blockchainsystemComponentRepository.MiningProcessComponent
import org.palladiosimulator.blockchainsystems.bscm.nodeallocation.NodeAllocation
import org.palladiosimulator.blockchainsystems.bscm.p2pnetwork.ConnectedSubgraphsNetworkTopology
import org.palladiosimulator.blockchainsystems.bscm.p2pnetwork.SubgraphSpecification
import org.palladiosimulator.blockchainsystems.core.system.abstractions.ResourcePowerCalculator

/**
 * This class calculates the global resource power of a connected subgraph network topology.
 *
 * @author Yannik Sproll, Davis Riedel
 */
class ConnectedSubgraphNetworkGlobalResourcePowerCalculator(
  connectedSubgraphsTopology: ConnectedSubgraphsNetworkTopology
) : ResourcePowerCalculator {
  private val globalResourcePower: Double = connectedSubgraphsTopology.subgraphs
    .sumOf { calculateResourcePowerOfSubgraph(it) }

  private fun calculateResourcePowerOfSubgraph(subgraph: SubgraphSpecification): Double {
    return subgraph.nodeTemplates
      .sumOf { getResourcePowerOfAllocation(it.allocation) }
  }

  override fun calculateGlobalResourcePower(): Double {
    return globalResourcePower
  }

  override fun getResourcePowerOfNode(nodeId: String): Double? {
    TODO("Not yet implemented")
  }

  private fun getResourcePowerOfAllocation(nodeAllocation: NodeAllocation): Double {
    return nodeAllocation
      .allocationContexts
      .filter { it.assemblyContext.encapsulatedComponent is MiningProcessComponent }
      .sumOf { it.resourceContainer.resourcePower }
  }
}